hypershift-operator/controllers/hostedcluster/internal/platform/agent/agent.go

```diff
@@ -45,13 +54,151 @@ func (p Agent) ReconcileCAPIInfraCR(ctx context.Context, c client.Client, create
+		Selector: &metav1.LabelSelector{
+			MatchLabels: map[string]string{
+				"control-plane": "controller-manager",
+			},
+		},
+		Template: corev1.PodTemplateSpec{
+			ObjectMeta: metav1.ObjectMeta{
+				Labels: map[string]string{
+					"control-plane": "controller-manager",
+				},
+			},
+			Spec: corev1.PodSpec{
+				TerminationGracePeriodSeconds: k8sutilspointer.Int64Ptr(10),
+				ServiceAccountName:            "cluster-api-provider-agent-controller-manager",
+				SecurityContext: &corev1.PodSecurityContext{
+					RunAsNonRoot: k8sutilspointer.BoolPtr(true),
+				},
+				Containers: []corev1.Container{
+					{
+						Name:  "manager",
+						Image: imageCAPA,
+						Env: []corev1.EnvVar{
+							{
+								Name: "MY_NAMESPACE",
+								ValueFrom: &corev1.EnvVarSource{
+									FieldRef: &corev1.ObjectFieldSelector{
+										FieldPath: "metadata.namespace",
+									},
+								},
+							},
+						},
+						Command: []string{"/manager"},
+						Args: []string{
+							"--namespace", "$(MY_NAMESPACE)",
+							"--health-probe-bind-address=:8081",
+							"--metrics-bind-address=127.0.0.1:8080",
+							"--leader-elect",
+							"--agent-namespace", hcluster.Spec.Platform.Agent.AgentNamespace,
+						},
+						LivenessProbe: &corev1.Probe{
+							ProbeHandler: corev1.ProbeHandler{
+								HTTPGet: &corev1.HTTPGetAction{
+									Path: "/healthz",
+									Port: intstr.FromInt(8081),
+								},
+							},
+							InitialDelaySeconds: 15,
+							PeriodSeconds:       20,
+						},
+						ReadinessProbe: &corev1.Probe{
+							ProbeHandler: corev1.ProbeHandler{
+								HTTPGet: &corev1.HTTPGetAction{
+									Path: "/readyz",
+									Port: intstr.FromInt(8081),
+								},
+							},
+							InitialDelaySeconds: 15,
+							PeriodSeconds:       20,
+						},
+						Resources: corev1.ResourceRequirements{
+							Limits: corev1.ResourceList{
+								corev1.ResourceCPU:    resource.MustParse("200m"),
+								corev1.ResourceMemory: resource.MustParse("100Mi"),
+							},
+							Requests: corev1.ResourceList{
+								corev1.ResourceCPU:    resource.MustParse("100m"),
+								corev1.ResourceMemory: resource.MustParse("20Mi"),
+							},
+						},
+					},
+					{
+						Name:  "kube-rbac-proxy",
+						Image: "gcr.io/kubebuilder/kube-rbac-proxy:v0.8.0",
+						Args: []string{
+							"--secure-listen-address=0.0.0.0:8443",
+							"--upstream=http://127.0.0.1:8080/",
+							"--logtostderr=true",
+							"--v=10",
+						},
+						Ports: []corev1.ContainerPort{
+							{
+								Name:          "https",
+								ContainerPort: 8443,
+								Protocol:      corev1.ProtocolTCP,
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+	return deploymentSpec, nil
 }
 
+//TODO add a new method to Platform interface?
 func (p Agent) ReconcileCredentials(ctx context.Context, c client.Client, createOrUpdate upsert.CreateOrUpdateFN,
 	hcluster *hyperv1.HostedCluster,
 	controlPlaneNamespace string) error {
+
+	role := &rbacv1.Role{
```

Comment:
I question the role itself should be owned by hypershift vs only the RoleBinding, thoughts?

Reply:
I prefer if we create the Role and RoleBinding together here since it reduces the number of manual steps and prerequisites the Agent platform already has. 
Since we said we assume the user is allowed to consume Agents from the given namespace and the reason we don't use a ClusterRole is that we want to lower the permissions we give the controller I think we can create the Role instead of letting the user worry about it.

Reply:
> I think we can create the Role instead of letting the user worry about it.

Who is the user there? I think there could be multiple users/personas.
As an infra admin I own the agents namespace and its rbac so I create the role with its curated perms.
As a service provider/SRE I give you a cluster that uses those perms audited by you.
As a cluster operator I admin the resulting guest cluster.
As a dev I have access to run my workload within the guest cluster.

Having this outside of hypershift enables a persona different from the cluster provider to authorise agents to be consumed by a particular cluster. It decouples perms and let a cluster consumer to own their infra perms?

Thoughts? @alvaroaleman @csrwng @ironcladlou @sjenning 

We can always revisit this later anyways